### ISSUE TITLE
Slow startup on Wayland (sway)

### ISSUE THREAD

**Issue by Leon-Plickat** [jwilm/alacritty#3045]
_Which operating system does the issue occur on?_
Linux, Arch based
_If on linux, are you using X11 or Wayland?_
Wayland, i915

The distribution I am on recently updated alacritty to version 0.4.0 and ever since then the startup is considerably slower than previously. It used to be near instantanious, now there is a noticeable delay of almost a second before the window appears. 

I tried building alacritty from source (version 0.4.1-dev), which is only marginally faster than the version I got from the repos.

**Comment by mzimann** [comment]
I can confirm this. I'm using the same configuration.

**Comment by cole-h** [comment]
Running a release-compiled version of master at https://github.com/jwilm/alacritty/commit/b9c513bd27f26271c51ff32fed38ce99abfc7e0d, I get the following startup times:

(I use `fish`, so if you wish to run the following on {z,ba}sh, the syntax is different)

```fish
for i in (seq 0 10); time target/release/alacritty -e echo; end
```

<details><summary>master</summary>
<p>

```
0.12 user 0.02 system 78% cpu 0.18s total
0.11 user 0.03 system 82% cpu 0.17s total
0.11 user 0.03 system 76% cpu 0.19s total
0.10 user 0.04 system 86% cpu 0.16s total
0.12 user 0.02 system 89% cpu 0.16s total
0.12 user 0.02 system 88% cpu 0.16s total
0.11 user 0.03 system 90% cpu 0.16s total
0.12 user 0.02 system 77% cpu 0.18s total
0.11 user 0.03 system 80% cpu 0.18s total
0.11 user 0.03 system 85% cpu 0.17s total
0.11 user 0.03 system 83% cpu 0.17s total
```
</p>
</details>

Checking out v0.4.0 and building (and running) a release version gives me these times:

<details><summary>v0.4.0</summary>
<p>

```
0.12 user 0.02 system 71% cpu 0.20s total
0.12 user 0.02 system 84% cpu 0.17s total
0.12 user 0.02 system 82% cpu 0.17s total
0.12 user 0.01 system 85% cpu 0.16s total
0.11 user 0.03 system 82% cpu 0.17s total
0.12 user 0.01 system 78% cpu 0.18s total
0.12 user 0.02 system 82% cpu 0.17s total
0.11 user 0.03 system 80% cpu 0.18s total
0.12 user 0.02 system 75% cpu 0.19s total
0.12 user 0.02 system 88% cpu 0.16s total
0.12 user 0.02 system 79% cpu 0.18s total
```
</p>
</details>

I would check v0.3.3, but my system hangs when starting the terminal for some (likely unrelated) reason.

Are your times different than mine when running the same (admittedly synthetic) test? To my eye, these are near-instantaneous and perfectly acceptable to me.

Is it possible you also updated sway/wlroots at the same time and the regression might have been introduced that way? I'm on sway `version 1.2-3334d11a (Nov 25 2019, branch 'master')` and wlroots `0.8.1.r50.g5cde3592-1` (using the nouveau driver, for what it's worth).

**Comment by Leon-Plickat** [comment]
In bash that would be

     for i in $(seq 0 10); do time alacritty -e echo; done 2>&1 | grep real

which gives me the following with v0.4.0-dev:

    real	0m0,354s
    real	0m0,332s
    real	0m0,365s
    real	0m0,368s
    real	0m0,359s
    real	0m0,329s
    real	0m0,318s
    real	0m0,315s
    real	0m0,366s
    real	0m0,312s

and for v0.3.3:


    real	0m0,173s
    real	0m0,132s
    real	0m0,164s
    real	0m0,112s
    real	0m0,108s
    real	0m0,098s
    real	0m0,096s
    real	0m0,110s
    real	0m0,097s
    real	0m0,114s
    real	0m0,095s

> Is it possible you also updated sway/wlroots at the same time and the regression might have been introduced that way?

That the problem is definitely not caused by a sway/wlroots update is the only thing I am sure about here: I have both of them build from source to get some features not yet in a release. Sway version is `sway version 1.2-1d3cbe9f (Nov 22 2019, branch 'master')`, wlroots is on version 0.8.1.

**Comment by chrisduerr** [comment]
That definitely looks like a significant increase in compile time, could someone bisect which commit exactly broke this?

**Comment by Leon-Plickat** [comment]
> could someone bisect which commit exactly broke this?

b0c6fdff763f7271506d26d7e768e6377fdc691b is still good. I can not pinpoint it further, because starting with 729eef0c933831bccfeac6a355bdb410787fbe5f up to somehwere around the version bump to 0.4.0 I am unable to build due to

    Unable to update https://github.com/chrisduerr/glutin#89aed1cb

**Comment by chrisduerr** [comment]
Oh yeah, since we have been building against git commits for a while this makes things significantly harder.

Though 729eef0 would be a good guess as to what might have caused it.

**Comment by frangio** [comment]
I'm seeing this as well. Is it possible that it's related to the changes around the mouse cursor? A significant (50%?) part of the strace logs from opening a new terminal are related to opening cursor files.

**Comment by kchibisov** [comment]
@frangio This is done by libwayland-cursor, I can't repro this issue myself on intel igpu/AMDGPU systems with different versions of sway. The startup is always fast for me.

**Comment by mzimann** [comment]
It's not possible to compile earlier version of Alacritty, because the commit https://github.com/chrisduerr/glutin/commit/89aed1cbfe38532c15d073e3b0822362cab3e044, which is referenced to in `Cargo.lock`, isn't included in any branch anymore (probably because of a force push). Thus, when the project is cloned this commit is missing.

In order to build it, I have forked the project and reconstructed the commit. My [master](https://github.com/mzimann/glutin/commits/master) branch points to that commit now. Then I have done the following:

```shell
$ cd ~/.cargo/git/db/glutin-f805f5db31188ce2/
$ git remote add origin https://github.com/mzimann/glutin.git
$ git fetch
```
This ensures that the commit is included in the cargo's local cache. Now, it's possible to build Alacritty.

For the future, it would be the easiest when in the repository of chrisduerr includes a branch which points to that commit. Then there shouldn't be a problem anymore with failing builds, due to the fact that this commit is missing.

I have bisected the commit and indeed the commit  https://github.com/jwilm/alacritty/commit/729eef0c933831bccfeac6a355bdb410787fbe5f is causing the problem.

**Comment by chrisduerr** [comment]
> It's not possible to compile earlier version of Alacritty, because the commit chrisduerr/glutin@89aed1c, which is referenced to in Cargo.lock, isn't included in any branch anymore (probably because of a force push).

> For the future, it would be the easiest when in the repository of chrisduerr includes a branch which points to that commit. Then there shouldn't be a problem anymore with failing builds, due to the fact that this commit is missing.

Yeah, that's my bad, sorry. Usually we don't want to use git depedencies, but unfortunately the depedency in question has not been updated in a very long time. If I knew that ahead of time, I definitely would have gone for a more proper approach.

**Comment by frangio** [comment]
I've noticed that the slowness is only present on the HiDPI display of my laptop, and not on the (non-HiDPI) external display.

**Comment by kchibisov** [comment]
Wait, how much displays do you have? Is there someone with 1 monitor?

**Comment by kchibisov** [comment]
Ok, I feel like I can repro the issue. You should have multiple monitors with mixed dpi, I guess. I didn't have 2 monitors to test my previous guess before (well it was proven on IRC that you should have mixed dpi), however now I have required setup to take a look myself.

**Comment by Leon-Plickat** [comment]
I have this issue with a single monitor, a 1920x1080 laptop screen with no scaling. Connecting different monitors with different resolutions and changing the scaling in no way affects the startup time.

**Comment by kchibisov** [comment]
Hm, I guess you're right here. It seems like there's something when startup could degrade, and having more then one monitor can increase a chance of it happening. So I can repro your issue with debug version of alacritty on both monitors, but only on one of them in release mode. Also I've noticed, that startup time on one of them vary in release, so it's either 150ms or 300ms+.

**Comment by kchibisov** [comment]
Could someone test #3130 ?

**Comment by Leon-Plickat** [comment]
I tested in both wayfire and sway. #3130 seems to solve the issue.  In both compositors, the average startup time drops to levels comparable to v0.3.3 (only about 2ms slower, which fits into the error one should expect for a test like this).

**Comment by chrisduerr** [comment]
Fixed in #3130.
